### PR TITLE
vector.direction fix and new methods

### PR DESCRIPTION
**Pull request by narrnika**
```vector.direction({x=0, y=0, z=0}, {x=0.5, y=0.5, z=0.5})```
return {x=4, y=4, z=4};
```vector.direction({x=-1, y=-1, z=-1}, {x=1, y=1, z=1})```
return {x=0.25, y=0.25, z=0.25}...

So it should be?
If not, then maybe it's better like this:

### CODE REVIEW

**Comment by paramat** [comment]
The current code looks wrong indeed.

**Comment by paramat** [comment]
I think your commit needs to return a unit vector.

**Comment by narrnika** [comment]
>  ...return a unit vector.

If I'm not mistaken, this is already doing ```vector.normalize```

Because the lua_api.txt contains an incomplete description, I was guided by the http://dev.minetest.net/position : 

>  	direction from p1 to p2, maximal values are: {x=-1,y=-1,z=-1} and {x=1,y=1,z=1}

**Comment by narrnika** [comment]
Although there is only one argument used ...
Perhaps you are right..

**Comment by paramat** [comment]
I see, it intends to limit the direction to within a unit-box. However a direction vector is not very useful if its length varies with direction, so a unit vector is more useful. `vector.normalize` uses a single vector argument, so `direction` can simply create a vector between p1 and p2 and feed it into `normalize`.

**Comment by narrnika** [comment]
Ok, then tomorrow I'll make a new requester.
This must somehow be removed?

**Comment by paramat** [comment]
You can simply update this PR.

**Comment by paramat** [comment]
Ah i see you're not using branches. If it's too difficult for you a new PR is ok.

**Comment by HybridDog** [comment]
https://forum.minetest.net/viewtopic.php?p=222011#p222011

**Comment by narrnika** [comment]
changed: new(), normalize(), direction();
added: is_nil(), ceil(), sum().

File: builtin/common/vector.lua [hunk 1]
```diff
@@ -5,9 +5,10 @@ function vector.new(a, b, c)
 	if type(a) == "table" then
 		assert(a.x and a.y and a.z, "Invalid vector passed to vector.new()")
 		return {x=a.x, y=a.y, z=a.z}
-	elseif a then
-		assert(b and c, "Invalid arguments for vector.new()")
+	elseif a and b and c then
 		return {x=a, y=b, z=c}
+	elseif a then
+		return {x=a, y=a, z=a}
```
Comment (paramat) [inline builtin/common/vector.lua]: spaces around '=' here and line 7

File: builtin/common/vector.lua [hunk 2]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
...
 
 function vector.normalize(v)
-	local len = vector.length(v)
-	if len == 0 then
-		return {x=0, y=0, z=0}
-	else
-		return vector.divide(v, len)
-	end
+	if vector.is_nil(v) then return v end
+	local inv_len = 1/vector.length(v)
```
Comment (paramat) [inline builtin/common/vector.lua]: spaces around '/' and '*' below

File: builtin/common/vector.lua [hunk 3]
```diff
@@ -63,83 +79,96 @@ function vector.distance(a, b)
...
-	local z_raw = pos2.z - pos1.z
-	local x_abs = math.abs(x_raw)
-	local y_abs = math.abs(y_raw)
-	local z_abs = math.abs(z_raw)
-	if x_abs >= y_abs and
-	   x_abs >= z_abs then
-		y_raw = y_raw * (1 / x_abs)
-		z_raw = z_raw * (1 / x_abs)
-		x_raw = x_raw / x_abs
-	end
```
Comment (sofar) [inline builtin/common/vector.lua]: I think the original bug was that this should have been an `elseif` ... `elseif` section

File: builtin/common/vector.lua [hunk 4]
```diff
@@ -4,12 +4,13 @@ vector = {}
 function vector.new(a, b, c)
 	if type(a) == "table" then
 		assert(a.x and a.y and a.z, "Invalid vector passed to vector.new()")
-		return {x=a.x, y=a.y, z=a.z}
+		return {x = a.x, y = a.y, z = a.z}
+	elseif a and b and c then
```
Comment (sofar) [inline builtin/common/vector.lua]: Documentation update needed?

File: builtin/common/vector.lua [hunk 5]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
 	       a.z == b.z
 end
 
+function vector.is_nil(a)
```
Comment (sofar) [inline builtin/common/vector.lua]: no, a `nil` vector is the wrong word here I think

`is_zero` or `is_zero_vector` is probably the right name. I prefer `is_zero_vector` myself.

docs needed.

File: builtin/common/vector.lua [hunk 6]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
...
+	if vector.is_nil(v) then return v end
+	local inv_len = 1 / vector.length(v)
+	return {
+		x = v.x * inv_len,
+		y = v.y * inv_len,
+		z = v.z * inv_len,
+	}
+end
+
+function vector.ceil(v)
```
Comment (sofar) [inline builtin/common/vector.lua]: docs needed.

File: builtin/common/vector.lua [hunk 7]
```diff
@@ -63,83 +79,96 @@ function vector.distance(a, b)
...
+		return {
+			x = a.x / b,
 			y = a.y / b,
-			z = a.z / b}
+			z = a.z / b,
+		}
 	end
 end
 
+function vector.sum(vs)
```
Comment (sofar) [inline builtin/common/vector.lua]: docs needed.

File: builtin/common/vector.lua [hunk 8]
```diff
@@ -63,83 +79,96 @@ function vector.distance(a, b)
...
-		{x = math.max(a.x, b.x), y = math.max(a.y, b.y), z = math.max(a.z, b.z)}
+	return {
+		x = math.min(a.x, b.x),
+		y = math.min(a.y, b.y),
+		z = math.min(a.z, b.z),
+	}, {
+		x = math.max(a.x, b.x),
+		y = math.max(a.y, b.y),
+		z = math.max(a.z, b.z),
+	}
```
Comment (sofar) [inline builtin/common/vector.lua]: I should also really add a bunch of unit tests here.

File: builtin/common/vector.lua (hunk 5 above)
Comment (SmallJoker) [inline builtin/common/vector.lua]: Since "vector" is already in the table name, `is_zero` is sufficient.

**Comment by narrnika** [comment]
Replaced is_nil by is_zero.
The documentation is really needed, but it will be better if someone else writes it - my English is not good enough.

**Comment by paramat** [comment]
Ideally this should be separated into 3 commits: one for code style cleanup, one for fixing 'direction' and one for the new functions. There's too much here to squash into 1 commit and some devs may want the cleanup and the fix without the new functions.

**Comment by rubenwardy** [comment]
This is the perfect piece of code to unit test

File: builtin/common/vector.lua (hunk 4 above)
Comment (rubenwardy) [inline builtin/common/vector.lua]: There's a problem here if a mod has a bug, and it calls:

```
vector.new(1, null, 3)
```

so this isn't a good change

File: builtin/common/vector.lua [hunk 9]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
...
 end
 
 function vector.normalize(v)
-	local len = vector.length(v)
-	if len == 0 then
-		return {x=0, y=0, z=0}
-	else
-		return vector.divide(v, len)
-	end
+	if vector.is_nil(v) then return v end
```
Comment (rubenwardy) [inline builtin/common/vector.lua]: don't put the body of if-statements on one line

File: builtin/common/vector.lua [hunk 10]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
...
+	local inv_len = 1 / vector.length(v)
```
Comment (rubenwardy) [inline builtin/common/vector.lua]: possible divide by zero here, which is a change in behaviour

File: builtin/common/vector.lua [hunk 11]
```diff
@@ -63,83 +79,96 @@ function vector.distance(a, b)
...
-		y_raw = y_raw / y_abs
-	end
-	if z_abs >= y_abs and
-	   z_abs >= x_abs then
-		x_raw = x_raw * (1 / z_abs)
-		y_raw = y_raw * (1 / z_abs)
-		z_raw = z_raw / z_abs
-	end
-	return {x=x_raw, y=y_raw, z=z_raw}
+	return vector.normalize({
```
Comment (rubenwardy) [inline builtin/common/vector.lua]: this really should be unit tested to make sure

`vector.normalize(old_direction) == vector.direction`

File: builtin/common/vector.lua (hunk 7 above)
Comment (rubenwardy) [inline builtin/common/vector.lua]: I don't like how this is no longer a class method (as the first parameter is a list, not a vector table)

**Comment by paramat** [comment]
@narrnika needs work, still interested?

Maybe the bugfix can be split off and adopted by someone.

File: builtin/common/vector.lua (hunk 10 above)
Comment (paramat) [reply builtin/common/vector.lua]: But `vector.is_nil()` avoids a zero length vector?

File: builtin/common/vector.lua (hunk 9 above)
Comment (paramat) [reply builtin/common/vector.lua]: Should be `vector.is_zero()`?

File: builtin/common/vector.lua [hunk 12]
```diff
@@ -18,40 +19,55 @@ function vector.equals(a, b)
...
+	return a.x == 0 and
+	       a.y == 0 and
+	       a.z == 0
+end
+
 function vector.length(v)
 	return math.hypot(v.x, math.hypot(v.y, v.z))
 end
 
 function vector.normalize(v)
```
Comment (paramat) [inline builtin/common/vector.lua]: Not sure why this needs changing.

**Comment by paramat** [comment]
I think the bugfix for `direction` should be split off and merged, i'm willing to do that.
The rest of these changes have issues.

**Comment by paramat** [comment]
PR for bugfix part of this #6801

**Comment by paramat** [comment]
#6801 merged which is the important part.
The rest has issues, seems low priority and author is absent, closing.

**Pull request closed without merging**